internal/interpreter/highlight.rs

```diff
@@ -3,22 +3,171 @@
     PropertyVisibility, RepeatedElementInfo,
 };
+use i_slint_core::component::ComponentVTable;
+use i_slint_core::item_tree::ItemWeak;
 use i_slint_core::items::ItemRc;
+use i_slint_core::lengths::LogicalPoint;
 use i_slint_core::model::{ModelRc, VecModel};
 use std::cell::RefCell;
 use std::path::PathBuf;
 use std::rc::Rc;
+use vtable::VRc;
 
 const HIGHLIGHT_PROP: &str = "$highlights";
+const CURRENT_ELEMENT_CALLBACK_PROP: &str = "$currentElementCallback";
+const DESIGN_MODE_PROP: &str = "$designMode";
+
+fn next_item(item: &ItemRc) -> ItemRc {
+    if let Some(s) = item.next_sibling() {
+        return next_item_down(&s);
+    }
+    if let Some(p) = item.parent_item() {
+        return p;
+    } else {
+        return next_item_down(item);
+    }
+}
+
+fn next_item_down(item: &ItemRc) -> ItemRc {
+    if let Some(child) = item.first_child() {
+        return next_item_down(&child);
+    } else {
+        return item.clone();
+    }
+}
+
+fn find_item(item: &ItemRc, position: &LogicalPoint) -> ItemRc {
+    let mut next = item.clone();
+    loop {
+        next = next_item(&next);
+
+        if next == *item {
+            return next;
+        }
+        let offset = next.map_to_window(LogicalPoint::default());
+        let geometry = next.geometry().translate(offset.to_vector());
+
+        if geometry.contains(*position) {
+            return next;
+        }
+    }
+}
+
+fn element_providing_item(component: &DynamicComponentVRc, index: usize) -> Option<ElementRc> {
+    generativity::make_guard!(guard);
+    let c = component.unerase(guard);
+
+    return c.description().original_elements.get(index).cloned();
+}
+
+fn map_offset_to_line(
+    source_file: Option<SourceFile>,
+    start_offset: usize,
+    end_offset: usize,
+) -> (String, u32, u32, u32, u32) {
+    if let Some(sf) = source_file {
+        let file_name = sf.path().to_string_lossy().to_string();
+        let (start_line, start_column) = sf.line_column(start_offset);
+        let (end_line, end_column) = sf.line_column(end_offset);
+        return (
+            file_name,
+            start_line as u32,
+            start_column as u32,
+            end_line as u32,
+            end_column as u32,
+        );
+    } else {
+        return (String::new(), 0, 0, 0, 0);
+    }
+}
+
+struct DesignModeState {
+    pub current_item: Option<ItemWeak>,
+}
+
+pub fn set_current_element_information_callback(
+    component_instance: &DynamicComponentVRc,
+    callback: Box<dyn Fn(String, u32, u32, u32, u32) -> ()>,
+) {
+    let weak_component = VRc::downgrade(component_instance);
+
+    generativity::make_guard!(guard);
+    let c = component_instance.unerase(guard);
+
+    let _ = c.description().set_callback_handler(
+        c.borrow(),
+        CURRENT_ELEMENT_CALLBACK_PROP,
+        Box::new(move |values: &[Value]| -> Value {
+            static mut STATE: DesignModeState = DesignModeState { current_item: None };
```

Comment:
don't use static mut,  this is NOT safe at all.   It is going to be shared if there are more instance.

Here you may want to move a `RefCell<DesignModeState>`  to the closure.